controllers/glanceapi_controller.go

```diff
@@ -375,6 +419,37 @@ func (r *GlanceAPIReconciler) reconcileNormal(ctx context.Context, instance *gla
 	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.InputReadyMessage)
 	// run check OpenStack secret - end
 
+	// Get Glance Top level CR and fetch ExtraVolumes if defined
+	parentGlance := &glancev1.Glance{}
+
+	err = r.Client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      glance.GetOwningGlanceName(instance),
+			Namespace: instance.Namespace,
+		},
+		parentGlance)
+
+	if err != nil {
+		if !k8s_errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, "Reconciling Service: Glance CR not present")
+	}
+
+	// if the top-level resource (parentGlance) is empty, an empty array is
+	// assigned to instance.ExtraMounts
+	instance.ExtraMounts = parentGlance.Spec.ExtraMounts
```

Comment:
I consider this as hack. I think if data needs to flow from Glance to GlanceAPI that should flow through the GlanceAPI.Spec fields. I know that we try to reduce the CRD size but I think this step went too far in that regard. This change makes the connection between Glance and GlanceAPI unnecessarily bidirectional and complicates the design. This breaks the rule that the Spec defines the expected state of the resource, as it is now the Spec and the Parent's Spec defines it together.

Reply:
In theory as long as the pattern is well documented and the picture is clear it's not an hack. We store the hash of the data structure (as it would be an `input` for the sub-resource) and we react here in case it changes: the difference here is that instead being an external resource (e.g. the `TransportURL`), it's the parent.
This could probably be a pattern (unless a different strategy is known) to inject data from the top-level to the sub-resources, without exposing any parameter that could be changed in the CR.
However, the main point here is to figure out if I'm really breaking the rules or creating an anti-pattern: in that case I guess we won't be able to trim the CRD size and find a strategy at `meta-operator` level.
@dprince @abays @stuggi @Akrog thoughts on this?

Reply:
By using this pattern the GlanceAPI.Spec can be empty. Every field populated there is populated from the parent Glance instance. So instead of populating it, the GlanceAPI controller could read it out from it's parent and live with an empty Spec. I guess that would be super confusing.

TransportURL is different as GlanceAPI references a TransportURL instance by its name in a field in GlanceAPI.Spec. I think we discussed before the option to move ExtraMounts as a CRD and reference it by name in GlanceAPI.Spec. That would look a lot better to me as the GlanceAPI.Spec would still describe the whole expected state of GlanceAPI.

Reply:
Another option is to have ExtraMounts in GlanceAPI.Spec but does not include GlanceAPISpec directly into Glance.Spec as that triplicates the definition of ExtraMounts in Glance, once in Glance.ExtraMounts once in Glance.GlanceAPIExternal.ExtraMounts and once in Glance.GlanceAPIInternal.ExtraMounts. 

Instead create a GlanceAPITempalate that only has fields that are not directly populated from Glance.Spec. Then when the Glance controller creates GlanceAPI it creates it from GlanceAPITemplate + Glance.Spec.

As an example you can see how Nova.Spec does not include a full NovaAPISpec just a NovaAPITemplate.

Reply:
> I think I can't do what has been done for the `TransportURL` because of the nature of the data structure and the places where we want it to be injected and modified. I think also that the `GlanceAPITemplate` in the top level CR would be a clean approach to inject data as long as they're not duplicated in the sub-level resources. I see that basically a `NovaAPITemplate` is defined in [1], but I think the key point is having `NovaServiceBase` [2] that is included in `NovaAPISpec` [3] and in theory is not rendered in the CRD (the same way I did with `ExtraMounts`) because we have:
> 
> ```
> NovaServiceBase `json:",inline"`
> ```

The `inline` only spares a single named field layer in the CRD but all the content of NovaServiceBase is copied directly into NovaAPISpec: https://github.com/openstack-k8s-operators/nova-operator/blob/master/config/crd/bases/nova.openstack.org_nova.yaml#L54-L64 

> 
> Finally, you're building the whole spec in the `nova_controller` [4], and in the generated `novaAPI` the Spec is assigned:
> 
> ```
> 	op, err := controllerutil.CreateOrPatch(ctx, r.Client, api, func() error {
> 		api.Spec = apiSpec
> 
> 		err := controllerutil.SetControllerReference(instance, api, r.Scheme)
> 		if err != nil {
> 			return err
> 		}
> 
> 		return nil
> 	})
> ```
> 
> This should be enough to assign hidden fields to the sub-resource and avoid "watching" the parent resource and break the "top-down" approach, right?

There is no such thing as hidden field. If a CRD need a field filled then that field needs to show in in the CRD. What we can spare is how many times we mention the same field definition.

If the Glance.Spec.ExtraMounts is the only field that is written by the end user and the GlanceExternalAPI.Spec.ExtraMounts always holds a copy of the value of Glance.Spec.ExtraMounts then you don't need to add the ExtraMounts field in Glance.Spec.GlanceExternalAPI only in GlanceAPI.Spec. 

> 
> I'll create a separate patch where I try the approach suggested, so we can review and choose the way that works best. I'd like to stay consistent across the operators, so if that is the pattern, I'll try that. I'm also wondering if we can benefit of this approach to reduce the size of the `openstack-operator`, but that's another story I guess.
> 
> [1] https://github.com/openstack-k8s-operators/nova-operator/blob/master/api/v1beta1/novaapi_types.go#L34 [2] https://github.com/openstack-k8s-operators/nova-operator/blob/master/api/v1beta1/common_types.go#L26 [3] https://github.com/openstack-k8s-operators/nova-operator/blob/master/api/v1beta1/novaapi_types.go#L133 [4] https://github.com/openstack-k8s-operators/nova-operator/blob/master/controllers/nova_controller.go#L713

Reply:
> I agree that we need to be careful to have bidirectional dependencies on the CRDs/controllers. I am wondering if it would be cleaner and consistent if we follow the same pattern as with networkAttachments or transporturl and have a dedicated CRD for the extraMounts a user can create. The extraMounts, which has the CRD kind as part of the propagation could then be passed in
> 
>     * either pass as a name list into the service operators by the openstack-operator
> 
>     * or the extraMounts controller could set an annotation on the service CRD to reference which one to use.
> 

I would not prefer breaking the existing pattern because it's already in all the operators: in addition, even though I know that everything is driven by the `openstack-operator`, I still want to keep the standalone use case for development reasons, hence having the ability to explicitly set `extraVolumes` in the top-level CR is a requirement.
I think that pattern of the shared CR is good, but not applies to this use case. 
> 
> Just two ideas, there might be more. With this I think the service CRDs could get rid of the big extraMount type def as part of their CRD.

As @gibizer suggested, I think an additional idea that keeps the current approach unchanged is implemented in #191.
That's enough to not repeat the same `struct` in the top-level CRD (where it appears only once, removing the references from the sub-resources).
It will also cut a lot on both `Cinder` and `Manila`, where the struct is repeated not only for the API but for other services (it's a x4 in case of cinder, a x3 in case of Manila).